src/VisualStudio/Core/Def/ProjectSystem/OpenTextBufferProvider.cs

```diff
@@ -0,0 +1,316 @@
+
+        private uint _runningDocumentTableEventsCookie;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public OpenTextBufferProvider(
+            IThreadingContext threadingContext,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _foregroundAffinitization = new ForegroundThreadAffinitizedObject(threadingContext, assertIsForeground: false);
+
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+
+            // The running document table since 16.0 has limited operations that can be done in a free threaded manner, specifically fetching the service and advising events.
+            // This is specifically guaranteed by the shell that those limited operations are safe and do not cause RPCs, and it's important we don't try to fetch the service
+            // via a helper that will "helpfully" try to jump to the UI thread.
+            // All other operations need to be done on the UI thread.
+#pragma warning disable RS0030 // Do not used banned APIs
+            var runningDocumentTable = serviceProvider.GetService<SVsRunningDocumentTable, IVsRunningDocumentTable>();
+            _runningDocumentTable = (IVsRunningDocumentTable4)runningDocumentTable;
+#pragma warning restore RS0030 // Do not used banned APIs
+            runningDocumentTable.AdviseRunningDocTableEvents(this, out _runningDocumentTableEventsCookie);
+
+            // We also need to check for any documents that might have been open before we subscribed. That we do have to do on the UI thread.
+            var listener = listenerProvider.GetListener(FeatureAttribute.Workspace);
+            var asyncToken = listener.BeginAsyncOperation(nameof(CheckForExistingOpenDocumentsAsync));
+            CheckForExistingOpenDocumentsAsync(threadingContext).CompletesAsyncOperation(asyncToken);
+        }
+
+        private void RaiseEventForEachListener(Action<IOpenTextBufferEventListener> action)
+        {
+            _foregroundAffinitization.AssertIsForeground();
+
+            foreach (var listener in _listeners)
+            {
+                try
+                {
+                    action(listener);
+                }
+                catch (Exception e) when (FatalError.ReportAndCatch(e, ErrorSeverity.Critical))
+                {
+                    // We'll catch the exception; this way if one listener is broken, we don't end up breaking other features that might no longer get events. Any exceptions would get caught by the
+                    // RunningDocumentTable itself which wouldn't report them in a useful way regardless.
```

Comment:
Previously we had a separate RDT listener for each separate component of Roslyn -- this probably would have meant an errant exception being thrown from one wouldn't have impacted others since it would have been swallowed by the shell. The catch here ensure we'll give better reporting than before, but also ensures that one thing listening and breaking stuff won't cause other things to break.